pkg/controller/volume/persistentvolume/pv_controller_base.go

```diff
@@ -346,33 +349,36 @@ func (ctrl *PersistentVolumeController) updateClaimMigrationAnnotations(ctx cont
 	return newClaim, nil
 }
 
-func (ctrl *PersistentVolumeController) updateVolumeMigrationAnnotations(ctx context.Context, volume *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+func (ctrl *PersistentVolumeController) updateVolumeMigrationAnnotationsAndFinalizers(ctx context.Context,
+	volume *v1.PersistentVolume) (*v1.PersistentVolume, error) {
 	volumeClone := volume.DeepCopy()
-	modified := updateMigrationAnnotations(ctrl.csiMigratedPluginManager, ctrl.translator, volumeClone.Annotations, false)
+	modified := updateMigrationAnnotationsAndFinalizers(ctrl.csiMigratedPluginManager, ctrl.translator,
+		volumeClone.Annotations, &volumeClone.Finalizers, false)
 	if !modified {
 		return volumeClone, nil
 	}
 	newVol, err := ctrl.kubeClient.CoreV1().PersistentVolumes().Update(ctx, volumeClone, metav1.UpdateOptions{})
 	if err != nil {
-		return nil, fmt.Errorf("persistent Volume Controller can't anneal migration annotations: %v", err)
+		return nil, fmt.Errorf("persistent Volume Controller can't anneal migration annotations or finalizer: %v", err)
 	}
 	_, err = ctrl.storeVolumeUpdate(newVol)
 	if err != nil {
-		return nil, fmt.Errorf("persistent Volume Controller can't anneal migration annotations: %v", err)
+		return nil, fmt.Errorf("persistent Volume Controller can't anneal migration annotations or finalizer: %v", err)
 	}
 	return newVol, nil
-
 }
 
-// updateMigrationAnnotations takes an Annotations map and checks for a
+// updateMigrationAnnotationsAndFinalizers takes an Annotations map and finalizers slice and checks for a
 // provisioner name depending if the annotation came from a PVC or not.
 // It will then add a "pv.kubernetes.io/migrated-to" annotation if migration with
 // the CSI driver name for that provisioner is "on" based on feature flags, it will also
-// remove the annotation if migration is "off" for that provisioner in rollback
-// scenarios. Returns true if the annotations map was modified and false otherwise.
+// remove the PV deletion protection finalizer and "pv.kubernetes.io/migrated-to" annotation
```

Comment:
@jsafrane @xing-yang I'm not clear why the finalizer feature is related to csi migration. Could you explain the reasoning?

We're trying to turn on gce migration in https://github.com/kubernetes/kubernetes/pull/104722 but the unit test is failing.

Reply:
If the finalizer feature is turned on, the finalizer will be added to all csi volumes including csi migrated volumes as well.  When the finalizer feature is turned off, we need to make sure the finalizers are deleted.

Reply:
I'm not clear why the logic to remove the finalizer is tied with the logic that determines if migration is enabled or not. I think it should be independent of any migration logic.

Reply:
@deepakkinni can you please fix the unit tests? They should set their own feature gate(s) and not depend on the current default in kube_features.go.

Reply:
Now I remember @deepakkinni said something about public holiday, I submitted https://github.com/kubernetes/kubernetes/pull/106376